fem/fe/fe_pyramid.hpp

```diff
@@ -0,0 +1,489 @@
+   mutable Vector      E_T_vtmp3;
+   mutable DenseMatrix E_T_mtmp1;
+   mutable DenseMatrix E_T_mtmp2;
+   mutable DenseMatrix V_Q_mtmp1;
+   mutable DenseMatrix V_Q_mtmp2;
+   mutable Vector      V_T_vtmp1;
+   mutable Vector      V_T_vtmp2;
+   mutable DenseMatrix V_T_mtmp1;
+   mutable Vector      VT_T_vtmp1;
+   mutable Vector      VT_T_vtmp2;
+   mutable DenseMatrix VT_T_mtmp1;
+   mutable DenseTensor VT_T_ttmp1;
+   mutable Vector      V_L_vtmp1;
+   mutable Vector      V_L_vtmp2;
+   mutable DenseMatrix V_L_mtmp1;
+   mutable DenseMatrix V_L_mtmp2;
+   mutable Vector      V_R_vtmp;
+   mutable DenseMatrix V_R_mtmp;
+#endif
+
+protected:
+   static constexpr real_t one  = 1.0;
+   static constexpr real_t zero = 0.0;
+   static constexpr real_t apex_tol = 1e-8;
+
+public:
+   FuentesPyramid() = default;
+
+   static bool CheckZ(real_t z) { return std::abs(z - 1.0) > apex_tol; }
+
+   /// Pyramid "Affine" Coordinates
+   static real_t lam1(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? (1.0 - x - z) * (1.0 - y - z) / (1.0 - z): 0.0; }
+   static real_t lam2(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? x * (1.0 - y - z) / (1.0 - z): 0.0; }
+   static real_t lam3(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? x * y / (1.0 - z): 0.0; }
+   static real_t lam4(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? (1.0 - x - z) * y / (1.0 - z): 0.0; }
+   static real_t lam5(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? z : 1.0; }
+
+   /// Gradients of the "Affine" Coordinates
+   static Vector grad_lam1(real_t x, real_t y, real_t z);
+   static Vector grad_lam2(real_t x, real_t y, real_t z);
+   static Vector grad_lam3(real_t x, real_t y, real_t z);
+   static Vector grad_lam4(real_t x, real_t y, real_t z);
+   static Vector grad_lam5(real_t x, real_t y, real_t z);
+
+   /// Two component vectors associated with edges touching the apex
+   static Vector lam15(real_t x, real_t y, real_t z)
+   { return Vector({lam1(x, y, z), lam5(x, y, z)}); }
+   static Vector lam25(real_t x, real_t y, real_t z)
+   { return Vector({lam2(x, y, z), lam5(x, y, z)}); }
+   static Vector lam35(real_t x, real_t y, real_t z)
+   { return Vector({lam3(x, y, z), lam5(x, y, z)}); }
+   static Vector lam45(real_t x, real_t y, real_t z)
+   { return Vector({lam4(x, y, z), lam5(x, y, z)}); }
+
+   /// Gradients of the above two component vectors
+   static DenseMatrix grad_lam15(real_t x, real_t y, real_t z);
+   static DenseMatrix grad_lam25(real_t x, real_t y, real_t z);
+   static DenseMatrix grad_lam35(real_t x, real_t y, real_t z);
+   static DenseMatrix grad_lam45(real_t x, real_t y, real_t z);
+
+   static Vector lam15_grad_lam15(real_t x, real_t y, real_t z);
+   static Vector lam25_grad_lam25(real_t x, real_t y, real_t z);
+   static Vector lam35_grad_lam35(real_t x, real_t y, real_t z);
+   static Vector lam45_grad_lam45(real_t x, real_t y, real_t z);
+
+   /// Three component vectors associated with triangular faces
```

Comment:
Are these explicit formulae for `s_0 \nabla s_1 \times \nabla s_2 + s_1 \nabla s_2 \times \nabla s_0 + s_2 \nabla s_0 \times \nabla s_1` ?

I tried to verify this guess, but there are so many terms that I think I would need to use Mathematica.